### PR TITLE
optimize bulk update/insert

### PR DESCRIPTION
**Pull request by NileshPant1999**

### CODE REVIEW

File: apps/mappings/tasks.py [hunk 1]
```diff
@@ -78,27 +78,10 @@ def auto_create_project_mappings(workspace_id):
...
-
-            mapping.source.auto_mapped = True
-            mapping.source.auto_created = True
-            mapping.source.save()
-
-            project_mappings.append(mapping)
-
-        return project_mappings
+        si_projects = upload_projects_to_fyle(workspace_id=workspace_id)
+        Mapping.bulk_create_mappings(fyle_categories, 'PROJECT', 'PROJECT', workspace_id)
```
Comment (NileshPant1999) [inline apps/mappings/tasks.py:84]: si_projects

File: apps/fyle/utils.py [hunk 2]
```diff
@@ -97,7 +97,7 @@ def _get_request(self, url, params):
             raise UnauthorizedClientError('Wrong client secret or/and refresh token', response.text)
 
         elif response.status_code == 404:
-            raise NotFoundClientError('Client ID doesn\'t exist', response.text)
+            raise NotFoundClientError('Client ID dosen\'t exist', response.text)
```
Comment (NileshPant1999) [inline apps/fyle/utils.py:100]:
```suggestion
            raise NotFoundClientError('Client ID doesn\'t exist', response.text)
```

File: apps/mappings/tasks.py (hunk 1 above)
Comment (NileshPant1999) [inline apps/mappings/tasks.py:84]:
```suggestion
        Mapping.bulk_create_mappings(si_projects, 'PROJECT', 'PROJECT', workspace_id)
```

File: apps/fyle/utils.py [hunk 3]
```diff
@@ -223,9 +225,10 @@ def sync_projects(self):
                 'source_id': project['id']
             })
 
-        project_attributes = ExpenseAttribute.bulk_upsert_expense_attributes(project_attributes, self.workspace_id)
+        ExpenseAttribute.bulk_create_or_update_expense_attributes(
+            project_attributes, "PROJECT", self.workspace_id)
```
Comment (ashwin1111) [inline apps/fyle/utils.py:229]: 'PROJECT'

File: apps/sage_intacct/utils.py [hunk 4]
```diff
@@ -53,10 +53,13 @@ def sync_accounts(self, workspace_id):
         general_settings: WorkspaceGeneralSettings = WorkspaceGeneralSettings.objects.filter(
             workspace_id=workspace_id).first()
 
-        account_attributes = []
+        account_attributes = {
+            "bank_account": [],
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:57]: 'account'

File: apps/sage_intacct/utils.py [hunk 5]
```diff
@@ -53,10 +53,13 @@ def sync_accounts(self, workspace_id):
...
+            "credit_card_account": []
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:58]: 'ccc_account'

File: apps/sage_intacct/utils.py [hunk 6]
```diff
@@ -53,10 +53,13 @@ def sync_accounts(self, workspace_id):
...
+        }
 
         for account in accounts:
-            account_attributes.append({
+            account_attributes['bank_account'].append({
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:62]: same as above

File: apps/sage_intacct/utils.py [hunk 7]
```diff
@@ -65,17 +68,19 @@ def sync_accounts(self, workspace_id):
             })
 
             if general_settings and general_settings.corporate_credit_card_expenses_object:
-                account_attributes.append({
+                account_attributes['credit_card_account'].append({
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:71]: same as above

File: apps/sage_intacct/utils.py [hunk 8]
```diff
@@ -185,9 +193,10 @@ def sync_projects(self):
                 'detail': detail
             })
 
-        account_attributes = DestinationAttribute.bulk_upsert_destination_attributes(
-            project_attributes, self.workspace_id)
-        return account_attributes
+        DestinationAttribute.bulk_create_or_update_destination_attributes(
+            project_attributes, 'PROJECTS', self.workspace_id, True)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:197]: 'PROJECT'

File: apps/sage_intacct/utils.py [hunk 9]
```diff
@@ -390,10 +402,11 @@ def sync_vendors(self, workspace_id: str):
                     'detail': detail
                 })
 
-        account_attributes = DestinationAttribute.bulk_upsert_destination_attributes(
-            vendor_attributes, self.workspace_id)
-        return account_attributes
-    
+        account_attributes = DestinationAttribute.bulk_create_or_update_destination_attributes(
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:405]: no use in storing them in a variable

Comment (ashwin1111) [reply apps/sage_intacct/utils.py:405]: also, we are appending `CHARGE_CARD_NUMBER` as well, current approach doesn't work

File: apps/sage_intacct/utils.py [hunk 10]
```diff
@@ -425,16 +438,16 @@ def post_vendor(self, vendor_name: str, email: str = None):
                 created_vendor = get_vendor['vendor'][0]
             else:
                 created_vendor = get_vendor['vendor']
-                
-        created_vendor = DestinationAttribute.bulk_upsert_destination_attributes([{
+
+        created_vendor = DestinationAttribute.bulk_create_or_update_destination_attributes({
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:442]: Shouldn't this be `create_or_update_destination_attributes`?

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

File: apps/sage_intacct/utils.py (hunk 9 above)
Comment (NileshPant1999) [reply apps/sage_intacct/utils.py:405]: ohh yes i didn't see that will update the logic

File: apps/sage_intacct/utils.py (hunk 10 above)
Comment (NileshPant1999) [reply apps/sage_intacct/utils.py:442]: sorry i forget to push the latest changes.

File: apps/mappings/tasks.py [hunk 11]
```diff
@@ -78,27 +97,12 @@ def auto_create_project_mappings(workspace_id):
...
-                destination_id=project.destination_id,
-                workspace_id=workspace_id
-            )
-
-            mapping.source.auto_mapped = True
-            mapping.source.auto_created = True
-            mapping.source.save()
-
-            project_mappings.append(mapping)
+        si_projects = upload_projects_to_fyle(workspace_id=workspace_id)
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:102]: upload_projects_to_fyle(workspace_id)

File: apps/mappings/tasks.py [hunk 12]
```diff
@@ -272,7 +279,7 @@ def async_auto_map_charge_card_account(workspace_id: int):
 def schedule_auto_map_charge_card_employees(workspace_id: int):
     general_settings = WorkspaceGeneralSettings.objects.get(workspace_id=workspace_id)
 
-    if general_settings.auto_map_employees and general_settings.corporate_credit_card_expenses_object:
+    if general_settings.auto_map_employees and general_settings.corporate_credit_card_expenses_object=='CHARGE_CARD_TRANSACTION':
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:282]: ```general_settings.corporate_credit_card_expenses_object == 'CHARGE_CARD_TRANSACTION'```

File: apps/mappings/tasks.py [hunk 13]
```diff
@@ -335,12 +342,19 @@ def upload_categories_to_fyle(workspace_id: int, reimbursable_expenses_object: s
...
+        si_attributes: List[DestinationAttribute] = DestinationAttribute.objects.filter(
+            workspace_id=workspace_id, attribute_type='EXPENSE_TYPE'
+        )
+
     else:
         si_connection.sync_accounts()
         si_attributes: List[DestinationAttribute] = DestinationAttribute.objects.filter(
             workspace_id=workspace_id, attribute_type='ACCOUNT'
         ).all()
+
```
Comment (ashwin1111) [inline apps/mappings/tasks.py:355]: rm

File: apps/sage_intacct/utils.py [hunk 14]
```diff
@@ -48,10 +48,13 @@ def sync_accounts(self):
         """
         accounts = self.connection.accounts.get_all()
 
-        account_attributes = []
+        account_attributes = {
+            "account": [],
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:52]: 'account'

File: apps/sage_intacct/utils.py [hunk 15]
```diff
@@ -48,10 +48,13 @@ def sync_accounts(self):
...
+            "ccc_account": []
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:53]: 'ccc_account'

**Review by ashwin1111 — APPROVED** [review]

**Review by Sravanksk — APPROVED** [review]

File: requirements.txt [hunk 16]
```diff
@@ -8,7 +8,7 @@ django-request-logging==0.7.1
 django-rest-framework==0.1.0
 djangorestframework==3.11.0
 future==0.18.2
-fyle_accounting_mappings==0.12.0
+fyle_accounting_mappings==0.16.6
```
Comment (ashwin1111) [inline requirements.txt:11]: set it to `0.15.0`

**Review by ashwin1111 — CHANGES_REQUESTED** [review]

**Pull request merged**